src/client/node/nodeKwil.ts

```diff
@@ -44,17 +48,91 @@ export class NodeKwil extends Kwil<EnvironmentType.NODE> {
       return await this.callClient(actionBody);
     }
 
-    let tempCookie: string | undefined;
+    // Ensure auth mode is set
+    await this.ensureAuthenticationMode();
+
+    // Set Temporary Cookie
+    const tempCookie = this.setTemporaryCookie(actionBody);
+
+    // Build Message
+    const message = await this.buildMessage(actionBody, kwilSigner);
+
+    // kwil.call()
+    let response = await this.callClient(message);
+
+    // Reset Cookie
+    if (this.authMode === AuthenticationMode.OPEN) {
+      this.resetCookie(tempCookie);
+    }
+
+    // Handle Authentication if error
+    response = await this.handleAuthentication(response, message, actionBody, kwilSigner);
+
+    return response;
+  }
+
+  /**
+   * Check if authMode is already set, if not call healthModeCheckClient()
+   * healthModeCheckClient => RPC call to retrieve health of blockchain and kwild mode (PRIVATE or OPEN (PUBLIC))
+   *
+   */
+  private async ensureAuthenticationMode(): Promise<void> {
+    if (!this.authMode) {
+      const health = await this.healthModeCheckClient();
+      this.authMode = health.data?.mode;
+    }
+  }
 
-    if (actionBody.cookie) {
-      // set the temp cookie so we can reset it after the call
-      tempCookie = this.cookie;
+  /**
+   * set the temp cookie so we can reset it after the call
+   *
+   * @param actionBody
+   * @returns the temporary cookie to handle for Node
+   * @returns undefined if in PRIVATE mode
+   */
+  private setTemporaryCookie(actionBody: ActionBodyNode): string | undefined {
+    if (this.authMode === AuthenticationMode.OPEN && actionBody.cookie) {
+      const tempCookie = this.cookie;
       this.cookie = actionBody.cookie;
+      return tempCookie;
+    }
+    return undefined;
+  }
+
+  /**
+   * Resets the temporary if in PUBLIC mode
+   *
+   * @param tempCookie
+   */
+  private resetCookie(tempCookie?: string): void {
+    if (this.authMode === AuthenticationMode.OPEN && tempCookie) {
```

Comment:
For readability, I would call the `if (this.authMode === AuthenticationMode.OPEN && tempCookie)` from within `kwil.call` before you call `resetCookie`.

The reason being is that when I first read through `call`, it read like there would still be the logic error where an undefined cookie would be reset on every call. However, once I clicked into `resetCookie`, I realized that wasn't the case. I think it makes more sense to move the conditional logic before calling the method, and then leaving the reset cookie logic (i.e. `this.cookie=undefined`) within the `resetCookie` method.